File: kotlin/programmers/lv1/기사단원의 무기.kt
class Solution {
  fun solution(number: Int, limit: Int, power: Int): Int {
      var arr = IntArray(number + 1, { value -> getDivisorCount(value) });
      
      var total =  arr.reduce { acc: Int, cur: Int -> if (cur > limit) acc + power else acc + cur;
      };
      
      
      return total;
  }
  
  private fun getDivisorCount(value: Int): Int {
      var result: MutableSet<Int> = mutableSetOf()
      
      if (value === 0) {
          return 0;
      }
      
      var maximumValue = Math.ceil(Math.sqrt(value.toDouble())).toInt();
      
      for (i in 1..maximumValue) {
          var quotient = value / i;
          var remainder = value % i;
          
         if (remainder === 0) {
             result.add(i)
             result.add(quotient);
         }
      }
      
      return result.size;
  } 
}

File: javascript/problems/programmers/lv.1/기사단원의 무기.js
function solution(number, limit, power) {
  const arr = new Array(number + 1).fill(0);

  for (let i = 1; i <= number; i += 1) {
    for (let j = i; j <= number; j += i) {
      arr[j] += 1;
    }
  }

  return arr.reduce(
    (acc, cur, idx) => (idx ? acc + (cur > limit ? power : cur) : acc),
    0
  );
}

console.log(solution(5, 3, 2));
